### PR TITLE
Feature/grouping performance

### PR DESCRIPTION
**Pull request by ivakegg**
This reverts the reverted grouping code and adds some performance enhancement in the RunningQuery.

### CODE REVIEW

File: web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java [hunk 1]
```diff
@@ -196,13 +206,186 @@ public void setConnection(Connector connection) throws Exception {
...
+     * the calls to hasNext and next on the underlying iterator. They will be decremented once a result is acknowledged in the RunningQuery.next() loop. The
+     * running boolean will allow the graceful termination of this thread.
+     * 
+     * @param queue
+     * @param running
+     * @param hasNext
+     * @param gotNext
+     * @return running (with a value of false)
+     */
+    private Object getResultsThread(BlockingQueue<Object> queue, AtomicBoolean running, AtomicInteger hasNext, AtomicInteger gotNext) {
```
Comment (apmoriarty) [inline web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java]: `hasNext` and `gotNext` are class level atomic ints, I don't think they need to be passed in as method parameters

File: web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java [hunk 2]
```diff
@@ -61,11 +66,17 @@ public class RunningQuery extends AbstractRunningQuery implements Runnable {
...
     private TInfo traceInfo = null;
     private transient QueryMetricsBean queryMetrics = null;
-    private RunningQueryTiming timing = null;
+    private transient RunningQueryTiming timing = null;
     private ExecutorService executor = null;
     private volatile Future<Object> future = null;
+    private BlockingQueue<Object> resultsThreadQueue = new ArrayBlockingQueue<>(1);
+    private AtomicInteger hasNext = new AtomicInteger(0);
+    private AtomicInteger gotNext = new AtomicInteger(0);
+    private AtomicBoolean running = new AtomicBoolean(false);
```
Comment (apmoriarty) [inline web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java]: all of these can be marked `final`

**Review by apmoriarty — DISMISSED** [review]
LGTM.

File: web-services/query/src/main/java/datawave/webservice/query/runner/QueryExecutorBean.java [hunk 3]
```diff
@@ -1368,18 +1369,18 @@ private BaseQueryResponse _next(RunningQuery query, String queryId, Collection<S
...
             // - race condition, query expired while user called next
             throw new PreConditionFailedQueryException(DatawaveErrorCode.QUERY_TIMEOUT_OR_SERVER_ERROR, e, MessageFormat.format("id = {0}", queryId));
         }
         
         long pageNum = query.getLastPageNumber();
         
-        BaseQueryResponse response = query.getLogic().getTransformer(query.getSettings()).createResponse(resultList);
-        if (!resultList.getResults().isEmpty()) {
+        BaseQueryResponse response = query.getLogic().getTransformer(query.getSettings()).createResponse(resultsPage);
+        if (resultsPage.getStatus() != ResultsPage.Status.NONE) {
```
Comment (cmitchell) [inline web-services/query/src/main/java/datawave/webservice/query/runner/QueryExecutorBean.java:1383]: This makes more semantic sense than checking for empty

**Review by cmitchell — DISMISSED** [review]
I didn't see anything I didn't expect to see.  Hopefully this nips this ticket in the bud

File: web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java (hunk 1 above)
Comment (ivakegg) [reply web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java]: sure, no problem

File: web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java (hunk 2 above)
Comment (ivakegg) [reply web-services/query/src/main/java/datawave/webservice/query/runner/RunningQuery.java]: yup

**Pull request merged**